openfe/protocols/openmm_utils/system_creation.py

```diff
@@ -0,0 +1,236 @@
+
+    system_generator = SystemGenerator(
+        forcefields=forcefield_settings.forcefields,
+        small_molecule_forcefield=forcefield_settings.small_molecule_forcefield,
+        forcefield_kwargs=forcefield_kwargs,
+        nonperiodic_forcefield_kwargs=nonperiodic_kwargs,
+        periodic_forcefield_kwargs=periodic_kwargs,
+        cache=str(cache),
+    )
+
+    # Add a barostat if necessary
+    # TODO: move this to its own place where we can then handle membranes
+    # Note: this behaviour was broken pre 0.11.2 of openmmff
+    if has_solvent:
+        barostat = MonteCarloBarostat(
+            ensure_quantity(thermo_settings.pressure, 'openmm'),
+            ensure_quantity(thermo_settings.temperature, 'openmm'),
+        )
+        system_generator.barostat = barostat
+
+    return system_generator
+
+
+ModellerReturn = Tuple[app.Modeller, Dict[Component, npt.NDArray]]
+
+
+def get_omm_modeller(protein_comp: Optional[ProteinComponent],
+                     solvent_comp: Optional[SolventComponent],
+                     small_mols: Dict[Component, OFFMol],
+                     omm_forcefield : app.ForceField,
+                     solvent_settings : SolvationSettings) -> ModellerReturn:
+    """
+    Generate an OpenMM Modeller class based on a potential input ProteinComponent,
+    SolventComponent, and a set of small molecules.
+
+    Parameters
+    ----------
+    protein_comp : Optional[ProteinComponent]
+      Protein Component, if it exists.
+    solvent_comp : Optional[ProteinCompoinent]
+      Solvent Component, if it exists.
+    small_mols : Dict[Component, openff.toolkit.Molecule]
+      Dictionary of SmallMoleculeComponents and their associated
+      OpenFF Molecule.
+    omm_forcefield : app.ForceField
+      ForceField object for system.
+    solvent_settings : SolvationSettings
+      Solvation settings.
+
+    Returns
+    -------
+    system_modeller : app.Modeller
+      OpenMM Modeller object generated from ProteinComponent and
+      OpenFF Molecules.
+    component_resids : Dict[Component, npt.NDArray]
+      Dictionary of residue indices for each component in system.
+    """
+    component_resids = {}
+
+    def _add_small_mol(comp: Component, mol: OFFMol,
+                       system_modeller: app.Modeller,
+                       comp_resids: Dict[Component, npt.NDArray]):
+        """
+        Helper method to add OFFMol to an existing Modeller object and
+        update a dictionary tracking residue indices for each component.
+        """
+        omm_top = mol.to_topology().to_openmm()
+        system_modeller.add(
+            omm_top,
+            ensure_quantity(mol.conformers[0], 'openmm')
+        )
+
+        nres = omm_top.getNumResidues()
+        resids = [res.index for res in system_modeller.topology.residues()]
+        comp_resids[comp] = np.array(resids[-nres:])
+
+    # If there's a protein in the system, we add it first to the Modeller
+    if protein_comp is not None:
```

Comment:
there's some repetition here in how small molecules are added.  Instead could we do?

```python
m = app.Modeller(app.Topology(), [])

if protein_comp:
    m.add(protein_comp.to_openmm, etc)

for comp, mol in small_mol.items():
    _add_small_mol(comp, mol, m, component_resids)
```

Reply:
No because you can't create an empty Modeller class, it always needs initialising with _something_

Reply:
Possibly, I remember trying to create an empty Modeller instance and having a hard time, but we can try.